metaflow_extensions/plugins/project_environment.py

```diff
@@ -205,13 +205,12 @@ def bootstrap_commands(self, step_name):
         # Pre-install specified packages if the user have them
         # installed in their local python env
         pip_install = f"{self.executable(step_name)} -m pip install --quiet"
+        cmds += [f"{pip_install} --upgrade pip"]
         for pkg_spec in filter(None, map(get_pkg_spec, PREINSTALL_PKGS)):
             cmds += [
-                f"{pip_install} --upgrade pip",
                 f"{pip_install} {pkg_spec} 1> /dev/null",
             ]
 
         # Install flow project
-        cmds.append(f"{pip_install} pkg_self/.")
-
+        cmds.append(f"{pip_install} -e pkg_self/. || export PYTHONPATH=$PWD/pkg_self")
```

Comment:
@bishax this took me about a week to fix in production... there were two separate situations that I was in:

1. Running metaflow (with batch steps) locally, which required the `-e` as discussed earlier (to allow for non-client packages that aren't really meant to be installed, such as a daps)
2. Running metaflow from an `amazonlinux2` container. For some reason, batch steps launched from a container like this (locally, and also in production) have problems running the line `pip install pkg_self/.`. No idea why, and I tried *everything*. In the end I can only recommend the nuclear option of not installing the package at all, and instead just exporting the PYTHONPATH. Very shit option!

For info, the error message I get just says that `.egg_info` can't be found. All threads / docs / q&a online seem to suggest that it's related to updating `setuptools`, but I had previously tried that and other debugging, even breaking down the install, all to no avail.

Reply:
:slightly_frowning_face: Sorry it took so long, sounds like a right pain in the arse!

Do you have any of those links still around? And to confirm when you say "locally" you mean that running an `amazonlinux2` container on your laptop reproduces the issue?

Reply:
The "full" error message is simply:

```
ERROR: No .egg-info directory found in /tmp/pip-pip-egg-info-f55pfh_d'
```

and the rabbit hole of links start from:

* https://github.com/actions/setup-python/issues/165
* https://github.com/pypa/pip/issues/8760
* https://stackoverflow.com/questions/26091641/what-does-a-could-not-find-egg-info-directory-in-install-record-from-pip-mean
* https://setuptools.pypa.io/en/latest/deprecated/python_eggs.html

> And to confirm when you say "locally" you mean that running an amazonlinux2 container on your laptop reproduces the issue?

Yes that's right:

https://github.com/nestauk/ojd_daps/blob/dev/ojd_daps/config/metaflowtask/Dockerfile-base
https://github.com/nestauk/ojd_daps/blob/dev/ojd_daps/config/metaflowtask/Dockerfile